[WIP] Function for String cleaning

jeffin143 opened the pull request:
@lozhnikov , Adhering to the GSOC timeline, I would like to complete string cleaning functions, till june 12, I have kept it as standalone function, Would like your take on the API.

jeffin143 commented on src/mlpack/CMakeLists.txt:
```diff
@@ -88,6 +88,13 @@ foreach(incl_file ${INCLUDE_FILES})
+  COMMAND ${CMAKE_COMMAND} ARGS -E
+  copy ${CMAKE_CURRENT_SOURCE_DIR}/core/data/stopwords.txt
+       ${CMAKE_BINARY_DIR}/include/mlpack/core/data/
+)
```
Above addition to copy the stopword.txt file at require place !. Will remove the comment it is just for debugging puprose.

lozhnikov replied:
I think the user should pass the stop words to CLI. I suggest to remove the file from the repo at all.

jeffin143 replied:
@lozhnikov , @zoq , both of you suggested to remove the file and hence i removed it.

jeffin143 commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,122 @@
+    return boost::hash_range<const char*>(str.begin(), str.end());
+  }
+};
+ 
```
Now , Since standalone function, can't migrate the class above . Any other workaround.

jeffin143 commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,122 @@
+        {
+          // If found then erase it from string
+          str.erase(pos, tokenStr.length() + 1);
+        }
```
Above line is sketchy, since I don't know what length will be delimiter of .

rcurtin replied:
Perhaps it would be worthwhile to make the tokenizer also return the token it split on, so you can put it back in here?

jeffin143 commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,124 @@
+std::string stopwords =
+#include "mlpack/core/data/stopwords.txt"
+;
+
```
Linter failes here, `;`, I have done this to include the file.

zoq replied:
Are you playing to keep this in the final version? I thought it might be useful to allow the user to pass a stop word file?

jeffin143 replied:
I removed the file, since both of you suggested to have an api, where user passes the file.

jeffin143 commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,155 @@
+  {
+    // Not sure how to take input for tokenizer from cli.
+    //data::RemoveStopWords(input);
+  }
```
Not sure, how should i build the binding for removing stop words, The binding takes a `.csv` file as input name and just parse and cleans the column having string data.

Yashwants19 replied:
Here we can first load stopwords and than can pass delimeter from data::Strtok class.
May be like this:
 ```
  if (CLI::HasParam("stopwords"))
  {
    std::unordered_set<std::string>stopword;
    stopword.insert("the");
    stopword.insert("a");    
    stopword.insert("isn't");
    stopword.insert("is");
    data::string_cleaning obj;
    obj.RemoveStopWords(input, stopword, data::Strtok(delimeter));
  }
```
Here I m loading stopwords intentionally. But you can load the data with the help some stopword file.
Thank you May this is helpful to you.

jeffin143 replied:
> Here we can first load stopwords and than can pass delimeter from data::Strtok class.
> May be like this:
> 
> ```
>  if (CLI::HasParam("stopwords"))
>  {
>    std::unordered_set<std::string>stopword;
>    stopword.insert("the");
>    stopword.insert("a");    
>    stopword.insert("isn't");
>    stopword.insert("is");
>    data::string_cleaning obj;
>    obj.RemoveStopWords(input, stopword, data::Strtok(delimeter));
>  }
> ```
> 
> Here I m loading stopwords intentionally. But you can load the data with the help some stopword file.
> Thank you May this is helpful to you.

@Yashwants19 , the issues is not with passing of stopwords, the issue is - if you take a close look the function it implements custom tokenization, my question is how to take input for custom tokenization from cli or Python :)

rcurtin replied:
I would say it is okay to not have custom tokenization in the binding here.  Typically the bindings provide "simple" functionality that will cover most popular use cases, but for "complex" tasks the user might have to do it in C++, which is fine.

To @Yashwants19's point of stop words, I do think it would be useful to specify a file of stop words.  But the tricky thing is that this is a little awkward from Python, since right now to use this binding you would have to pass filenames---which is a bit weird since in Python one would expect the user to already have the stopwords as a loaded list of strings or something.

It's up to you if you want to handle this here, but one way around this would be to do the following: we can modify the documentation for `PARAM_VECTOR_STR_IN()` so that if the given vector has only one element and that element has `file:` as the first part, then the rest of it is a filename where we should load one string per line.  The way to do this would be to modify `src/mlpack/bindings/cli/get_param_impl.hpp` for the case where the type is `std::vector<std::string>`, and do the same for `src/mlpack/bindings/python/get_param_impl.hpp`.  In fact the function that checks if the vector contains only one element and then loads from file if needed can be put into some utility file in `src/mlpack/core/util/` or something and then just called from the `get_param_impl.hpp` file.  Anyway, I hope this makes sense; let me know if I can clarify the idea or anything.  And don't feel obligated to do it, but it could make the binding here cleaner to use.

zoq commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,115 @@
+ *
+ * @param input Vector of strings
+ */
+void LowerCase(std::vector<std::string>& input)
```
I'm wondering if uppercase is something that we like to support as well.

jeffin143 replied:
Added a functionality for uppercase

zoq commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,115 @@
+ *
+ * Definiation of function should be of type
+ * boost::string_view fn()(boost::string_view& str)
+ *
```
Looks like we could remove the empty line here.

jeffin143 replied:
Agreed.

zoq commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,115 @@
+ *
+ */
+template<typename TokenizerType>
+void RemoveStopWords(std::vector<std::string>& input, TokenizerType tokenizer)
```
Should this be ``const TokenizerType& tokenizer``?

jeffin143 replied:
Yeah.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,115 @@
+template<typename TokenizerType>
+void RemoveStopWords(std::vector<std::string>& input, TokenizerType tokenizer)
+{
+  boost::string_view strView(stopwords);
```
I suggest to accept the stop words as an `unordered_set`.

jeffin143 replied:
Sure.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,115 @@
+        if (pos != std::string::npos)
+        {
+          // If found then erase it from string
+          str.erase(pos, tokenStr.length() + 1);
```
I think it's faster to write the result to another string. Since `erase()` takes `O(N)` operations each time.

jeffin143 replied:
Done.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,157 @@
+PARAM_STRING_IN_REQ("preprocess_dataset", "File containing the preprocess "
+    "dataset.", "o");
+
+PARAM_INT_IN("lowercase", "convert to lowercase(1 to convert).", "l", 1);
```
I think the parameters below shouldn't have values. They just could be passed or otherwise.

jeffin143 replied:
I am not sure, how should i implement that.

lozhnikov replied:
Try to use `PARAM_FLAG` (e.g. `src/mlpack/methods/cf/cf_main.cpp`).

jeffin143 replied:
Done.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,157 @@
+  {
+    for (size_t j =0 ; j < dataset[i].size(); j++)
+    {
+      if (j == Column)
```
Could you elaborate a bit, what is the purpose of the conditional jump below?

jeffin143 replied:
Suppose our dataset is a csv file with the following data
```
10,100,Hello world,15
12,500,Machine Learning,20
```

Now we have to process strings, in the following dataset and hence i am taking an input from user for which of the above column i have to do string processing and then separating the column of numerical dataset into one vector and , the string part to the other.

So we would have follwing two vectors
```
10,100,15
12,500,20
```

```
Hello world
Machine Learning
```
I will do the functionalities on the second one, while remerging i have to mind that it should be placed at correct location and hence conditional jump.

lozhnikov replied:
I think it is better to treat all data as strings since strings can contain digits.

jeffin143 replied:
Umm, If you take a close look over the code , I have taken both as strings, I have just used two vectors, such that user may not always want to do processing at every column. 
For instance take a case :

```
1,10, Hello world.,45.6
3,0,Mlpack i's great?,53.6
```
Now for the above case , user might want to remove punctuation from string, but if we pass whole vector, than it will also transform 45.6 -> 456 and 53.6 ->536, along with other strings since `.` is considered as punctuation and that totally changes its meaning, hence i will take 1st, 2nd and 4th column in vector and 3rd in another , and pass the vector which has only 3rd column, to perform function and then merge them back. 

In my code, I have used the following two vectors :

```
std::vector<std::string> input; // Store data of a particular column to perform function upon
std::vector<std::vector<std::string>> dataset; // rest all columns
```

lozhnikov replied:
Thanks for the clarification. But what is the sense of splitting a string which has already been split by commas?

jeffin143 replied:
How should I divide it into two vectors then, I am just separating a particular column. Which I need to pass it to the function. While reading the rows, they are just a single string separated by `,`. And hence I am splitting it to store in`vector<vector<string>>`.


Or should I pass the whole dataset at once to the function for preprocessing.? Or you mean instead of story `vector<vector<string>>` store it in `vector<string>`

jeffin143 commented:
> I glanced over the PR. Sorry for the slow response. How do you think, maybe it is better to add a class that could handle all cleaning operations instead of introducing so many standalone functions?

Implemented a class

Yashwants19 commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,124 @@
+static void mlpackMain()
+{
+  // Parse command line options.
+  // Ectracting the filename
```
I think so It must be extracting.

Yashwants19 commented:
Hi @jeffin143 if you don't mine can you please explain more about custom tokenizer. Here what do you mean by custom tokenizer. :)

jeffin143 commented:
> Hi @jeffin143 if you don't mine can you please explain more about custom tokenizer.What do you mean by custom tokenizer here. :)

sure, 
So basically custom tokenization means user can pass their own definition instead of passing the `data::strtok(" ")` object. They can use lambda function or create their own functors.

example :
```cpp
  obj.RemoveStopWords(input, stopword, [](boost::string_view& str) {
       if (str.empty())
        return str;
      boost::string_view retval = str.substr(0, 1);
      str.remove_prefix(1);
      return retval;
  });
```

Yashwants19 commented:
For Different tokenizer function we can use:
```
PARAM_STRING_IN("tokenizer",  " -",  "- ");
const std::string tokenizer = CLI::GetParam<std::string>("tokenizer");
 if (CLI::HasParam("stopwords"))
 {
   std::unordered_set<std::string>stopword;
   stopword.insert("the");
   stopword.insert("a");    
   stopword.insert("isn't");
   stopword.insert("is");
   data::string_cleaning obj;
   if(tokenizer == "A") 
     obj.RemoveStopWords(input, stopword, data::A());
   else if (tokenizer == "B")  
     obj.RemoveStopWords(input, stopword, data::B());
 }
```
But for custom tokenization or create own functors its interest to see that how we can set those custom tokenization through(cli bindings). All the best for this part :)

rcurtin commented on src/mlpack/core/boost_backport/boost_backport_string_view.hpp:
```diff
@@ -0,0 +1,39 @@
+      return boost::hash_range<const char*>(str.begin(), str.end());
+    }
+  };
+}
```
Should this be only for a specific boost version?

lozhnikov replied:
Yes, it's fixed in #1814. Looks like this PR isn't based on #1814. Actually, the specialization appeared in Boost 1.69.

rcurtin commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,108 @@
+namespace mlpack {
+namespace data {
+
+class string_cleaning
```
Should we call this class `StringCleaning` to match the rest of the mlpack style?

rcurtin commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,108 @@
+  *
+  * @param input Vector of strings
+  */
+  void RemovePunctuation(std::vector<std::string>& input)
```
It may be worth considering adding `std::string punctuation` as a parameter with default value `"!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}-"`, so that the user can specify their own set of punctuation characters, if perhaps there are some they would like to keep.

Maybe it is better instead to pass a function here like `std::isalnum()` or something like that, where if the function returns true then the character is kept?  (Or something kind of like that.)

lozhnikov replied:
I think `std::ispunct()` suits better than `std::isalnum()`. I agree we could pass either a string or a function.

rcurtin commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,108 @@
+  */
+  template<typename TokenizerType>
+  void RemoveStopWords(std::vector<std::string>& input,
+                       std::unordered_set<std::string>stopwords,
```
You might consider using some kind of `std::map<>` or something and not just an unordered set, since you will be searching for each stopword.  A sorted data structure can make that search a lot faster.

lozhnikov replied:
I believe `std::unordered_set` is implemented using hash tables. So the access time is `O(1)`. And `std::map` is a kind if binary tree (usually the red-black one), the access time is `O(log N)`. Actually, the choice depends on the complexity oh the hash function. If the hash of a string is being calculated faster than `O(N)` then we definitely should use `std::unordered_set`. However I guess it depends on the length of the string. So, I would recommend to write a benchmark.

lozhnikov replied:
I think `stopwords` should be a const reference.

rcurtin commented on src/mlpack/core/data/tokenizer/strtok.hpp:
```diff
@@ -0,0 +1,66 @@
+  * A constructor to set deimiter.
+  * @param Delimiter A string which you want to use as delimiter
+  */
+  Strtok(std::string Delimiter)
```
It might be easier to use initialization lists here: `Strtok(const std::string& delimiter) : delimiter(delimiter) { }`.

rcurtin commented on src/mlpack/core/data/tokenizer/strtok.hpp:
```diff
@@ -0,0 +1,66 @@
+  /**
+  * A function object which take boost::string_view as input and
+  * return a boost::string_view (token).
+  * @param str A string to retieve token from.
```
`retrieve` :)

rcurtin commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,124 @@
+PARAM_STRING_IN_REQ("actual_dataset", "File containing the reference dataset.",
+    "t");
+PARAM_STRING_IN_REQ("preprocess_dataset", "File containing the preprocess "
+    "dataset.", "o");
```
It might be clearer to show that this is an output parameter by writing `File that will contain the preprocessed dataset` as the comment here.

rcurtin commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,124 @@
+PARAM_STRING_IN_REQ("preprocess_dataset", "File containing the preprocess "
+    "dataset.", "o");
+PARAM_STRING_IN_REQ("delimeter", "delimeter used to seperate Column in files"
+    "example '\\t' for '.tsv' and ',' for '.csv'.", "d");
```
It's probably worth having a default here of `""`, and if the parameter is empty, you can use the file extension to "guess", just like `data::Load()` does.  Also it should be spelled `delimiter`. :)

rcurtin commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,124 @@
+PARAM_FLAG("punctuation", "Remove punctuation.", "p");
+PARAM_FLAG("stopwords", "Remove stopwords.", "s");
+PARAM_INT_IN("dimension", "Column which contains the string data.(1 by "
+    "default)", "c", 1);
```
Maybe `PARAM_VECTOR_INT_IN` is better here?  It would allow the user to specify multiple dimensions containing string data.  It may also be worth considering an `auto` option here or something that will detect non-numeric dimensions automatically.

rcurtin commented on src/mlpack/tests/CMakeLists.txt:
```diff
@@ -99,6 +99,7 @@ add_executable(mlpack_test
   sparse_coding_test.cpp
   spill_tree_test.cpp
   split_data_test.cpp
+  string_utility_test.cpp
```
Do you want to also add `main_tests/preprocess_string_test.cpp` to test that the different options of `preprocess_string_test` seem to work correctly?

rcurtin commented on src/mlpack/tests/string_utility_test.cpp:
```diff
@@ -0,0 +1,84 @@
+{
+  std::vector<string> arr(2);
+  arr[0] = "isn't mlpack great?";
+  arr[1] = "2019 gsoc is great.";
```
:)

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,108 @@
+        if (stopwords.find(tokenStr) == stopwords.end())
+        {
+          // token is not a stop word add it;
+          copy = copy + " " + tokenStr;
```
I think `operator+=()` would work faster than `=` since it doesn't necessarily copy the data.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,108 @@
+      token = tokenizer(strView);
+      while (!token.empty())
+      {
+        tokenStr = std::string(token);
```
Probably it's faster to accept a `string_view` rather than convert each token into a string.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,108 @@
+    boost::string_view strView;
+    for (auto& str : input)
+    {
+      copy = "";
```
I suggest to use `clear()`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,124 @@
+  std::vector<std::string> temp;
+  std::vector<std::string> input;
+  std::vector<std::vector<std::string>> dataset;
+  const size_t Column = CLI::GetParam<int>("dimension");
```
According to the style guide each variable should be camel case starting with a lower character.

lozhnikov replied:
I think the user should be able to pass a set of columns as well.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,124 @@
+  const size_t Column = CLI::GetParam<int>("dimension");
+  while (std::getline(fin, line))
+  {
+    boost::split(temp, line, boost::is_any_of(std::string(1, delimeter[0])));
```
Why didn't you pass `delimiter` as is?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,124 @@
+  while (std::getline(fin, line))
+  {
+    boost::split(temp, line, boost::is_any_of(std::string(1, delimeter[0])));
+    input.push_back(temp[Column]);
```
You should check that `temp.size() > column`. Otherwise it could be a segmentation fault.

lozhnikov replied:
I suggest to reduce memory usage: actually, you needn't to store `temp[column]` twice.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+PARAM_STRING_IN_REQ("delimeter", "delimeter used to seperate Column in files"
+    "example '\\t' for '.tsv' and ',' for '.csv'.", "d");
+PARAM_STRING_IN("strDelimeter", "A set of chars that is used as delimeter to"
+    "tokenize the string dataset ", "D", " ");
```
`strDelimiter`: If I am not mistaken all cmd arguments should be lowercase. And I don't lake the name of the arguments, since it's not meaningful. I suggest `column_delimiter` for columns and `delimiter` for tokens.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+  const std::string filename = CLI::GetParam<std::string>("actual_dataset");
+  // This is very dangerous, Let's add a check tommorrow.
+  const std::string delimiter = CLI::GetParam<std::string>("delimeter");
+  std::vector<int> dimension = CLI::GetParam<std::vector<int>>("dimension");
```
Probably we should add some checks for the parameters.

lozhnikov replied:
And I think it's really cool to support some complex dimensions like `["1", "3", "5-8", "10", "15"]`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+  // File pointer
+  fstream fin;
+  // Open an existing file
+  fin.open(filename, ios::in);
```
We need to check that the file has been opened successfully.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+  std::sort(dimension.begin(), dimension.end());
+  // Extracting the Contents of file
+  // File pointer
+  fstream fin;
```
Why don't you use `ifstream`?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+  std::vector<std::string> temp;
+  std::vector<std::vector<std::string>> dataset;
+  // const size_t Column = CLI::GetParam<int>("dimension");
+  data::SplitByChar sobj(delimiter);
```
`sobj`: could you replace it by something meaningful?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+    dataset.push_back(temp);
+    temp.clear();
+  }
+  fin.close();
```
There is no need to close the file since it will be closed in the destructor.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+    // Not sure how to take input for tokenizer from cli.
+    if (!CLI::HasParam("stopwordsfile"))
+    {
+      throw std::runtime_error("Please provide a file for stopwords.");
```
Look at the other CLI files. There are some functions like: `RequireOnlyOnePassed()`, `RequireAtLeastOnePassed()`, `RequireParamValue()`, etc. I think you should use these functions for the checks.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+    {
+      throw std::runtime_error("Please provide a file for stopwords.");
+    }
+    fstream fin;
```
I though `fin` is already defined. And again I prefer `ifstream` here.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+    // Open an existing file
+    const std::string stopwordfilename = CLI::GetParam<std::string>
+        ("stopwordsfile");
+    fin.open(stopwordfilename, ios::in);
```
Has the file been opened successfully?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+        break;
+      }
+    }
+  }
```
I am not sure that I understand this loop. I thought you wanted to process some peculiar columns. Why do you verify row numbers?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,188 @@
+  const std::string filename2 = CLI::GetParam<std::string>("preprocess"
+      "_dataset");
+  fstream fout;
+  fout.open(filename2, ios::out | ios::trunc);
```
I suggest `ofstream`. You should verify that the file has been opened successfully.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,132 @@
+        token = tokenizer(strView);
+      }
+      boost::algorithm::trim_left(copy);
+      str = copy;
```
I suggest to use move assignment operator.

jeffin143 commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,132 @@
+        if (stopwords.find(token) == stopwords.end())
+        {
+          // token is not a stop word add it;
+          copy += " " + std::string(token);
```
> The function doesn't preserve spaces. I am not sure it's correct. Did you compare this behavior with other toolkits? For example does scikit-learn preserve spaces?

Scikit-learn, doesn't have a standalone function, The have it in their pipline implementation, I compared it with NLTK library, the output of this function is
`vector<vector<string>> dataset` , After removal of stopword, they are tokenizing it,

And hence A good option would be to add a overload function in our `StringEncoding` Class which accepts `vector<vector<string>>` as input and then do the required thing, 

other one could be to introduce space and make it `vector<string> dataset`.

lozhnikov replied:
Okay, in this case we are not bound to preserve delimiters.
I think we don't need the overload since it's easy to handle this in terms of tokenizers. It's enough to write a tokenizer that returns the whole string.

jeffin143 replied:
> I think we don't need the overload since it's easy to handle this in terms of tokenizers. It's enough to write a tokenizer that returns the whole string.

Not sure what does this mean, You mean a function which accepts a `vector<vector<string>>` and delimiters and output a vector<string> with merger all strings ?

lozhnikov replied:
No, the function still accepts `vector<string>` and a tokenizer. We needn't anything else. I mean we don't need to handle this special case.

jeffin143 replied:
So i need not change anything. Right ?

`copy += " " + std::string(token);`

lozhnikov replied:
Correct.

jeffin143 commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,250 @@
+      dataset[dataset.size()-1].push_back(word);
+    }
+    streamLine.str(std::string());
+  }
```
@lozhnikov , Could you help me here, while reading the lines from a file, it is just reading the first line correctly into stringstream, for subsquent line it is not working, I added cout statements and saw that the contents are being added but it is not reading from stringstream 

![Screenshot from 2019-07-20 22-47-27](https://user-images.githubusercontent.com/21286134/61582064-a4598700-ab43-11e9-873c-a34d5dd12684.png)

check the output, I have added functionality for `3-5` and other as same, but it is just for one row.

jeffin143 replied:
Ok, So issue is resolved, it is reading all rows :), The punctuation has been removed from all columns

![Screenshot from 2019-07-21 17-27-24](https://user-images.githubusercontent.com/21286134/61590980-d4556880-abdd-11e9-8d2b-c1835af21626.png)

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+    {
+      if (col < dimension.size())
+      {
+        if (dimension[col] == j)
```
It won't work if `dimension` contains unique values or if `dimension` begins with negative values.

jeffin143 replied:
Not sure, How will unique values break the code ?

lozhnikov replied:
I think it can lead to the condition `dimension[col] < j`. In this case the condition `dimention[col] == 1` will never happen.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+    }
+    ifstream stopwordfile;
+    // Open an existing file
+    const std::string stopwordfilename = CLI::GetParam<std::string>
```
I suggest to break the line after `=`.

jeffin143 replied:
Ok

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+    stopwordfile.open(stopwordfilename);
+    if (!stopwordfile.is_open())
+    {
+      throw std::runtime_error("Unable to open the file for stopwords.");
```
I think `Log::Fatal` suits much better than the exception.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+  fout.open(filename2, ios::trunc);
+  if (!fout.is_open())
+  {
+    throw std::runtime_error("Unable to open a file for writing output.");
```
`Log::Fatal`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+    for (size_t i = 0; i < input.size(); i++)
+      obj.RemovePunctuation(input[i]);
+  }
+  const std::string filename2 = CLI::GetParam<std::string>("preprocess"
```
I suggest to replace `filename2` by `outputFilename`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+      }
+      else
+      {
+        fout<<dataset[i][j]<<column_delimiter;
```
Looks like the last delimiter is redundant (you should check that `j+1 < dataset[i].size()`).

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+  }
+  size_t col = 0;
+  // Preparing the input dataset on which string manipulation has to be done.
+  std::vector<std::vector<std::string>> input(dimension.size());
```
It's hard to understand the difference between `dataset` and `input`. Could you find a bit more meaningful names?

jeffin143 commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,131 @@
+    boost::string_view strView;
+    for (auto& str : input)
+    {
+      copy.clear();
```
Thanks @lozhnikov , I believed that there would be performance lag while creating the variable inside the loop, as in `for(){ int a}`, every iteration the variable will be created, Being a sport programmer I always believed That we should declare it at first and then use, but I researched a bit and found [this](https://stackoverflow.com/a/7959658/9559556). So i guess I will go with your suggestion.

Thanks

lozhnikov replied:
Hmm.. Perhaps you were right.
I thought it doesn't influence on the performance since it doesn't increase the number of heap allocations because at the end of each iteration the string is moved to the container. However, I didn't pay attention to the allocations performed by `operator+=`. It's quite interesting I think we should make the benchmark.

jeffin143 commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,131 @@
+  * bool fn(const char& c)
+  */
+  template<typename ReturnType>
+  void RemoveChar(std::vector<std::string>& input, const ReturnType fun)
```
Done.

jeffin143 commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,131 @@
+    for (auto& str : input)
+    {
+      boost::algorithm::to_lower(str);
+    }
```
Sure.

jeffin143 commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,131 @@
+        if (stopwords.find(token) == stopwords.end())
+        {
+          // token is not a stop word add it;
+          copy += " " + std::string(token);
```
Added a check.

jeffin143 commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+      }
+  }
+  // Handling Dimension vector
+  std::vector<std::string> temp_dimension = CLI::GetParam<std::vector<
```
Ok

jeffin143 commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+    if ( found != string::npos)
+    {
+      // Has a range include, something of type a-b.
+      columnstartindex = std::stoi(temp_dimension[i].substr(0, found));
```
Please take a look

lozhnikov replied:
At the first glance looks good except `Log::Fatal << e.what() << std::endl;`. Unfortunately `std::exception` doesn't describe the reason well.

jeffin143 commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+  }
+  std::cout<<"\n";
+  // Sorting neccessary
+  std::sort(dimension.begin(), dimension.end());
```
Instead of `std::array` made, `std::set`, arranged and unique, 

Again `nlogn + n`

jeffin143 commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,248 @@
+  fin.open(filename);
+  if (!fin.is_open())
+  {
+    throw std::runtime_error("Unable to open input file");
```
Done, Made changed.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,131 @@
+          result += " " + std::string(token);
+        }
+        token = tokenizer(strView);
+      }
```
I think we can reduce the number of lines e.g.
```c++
token = tokenizer(strView);
if (!token.empty() && stopwords.find(token) == stopwords.end())
  result += std::string(token);

if (!token.empty())
  token = tokenizer(strView);

while(!token.empty())
{
  if (stopwords.find(token) == stopwords.end())
    result += " " + std::string(token);

  token = tokenizer(strView);
}
```

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+      }
+      catch (const std::exception& e)
+      {
+        Log::Fatal << e.what() << std::endl;
```
Usually `e.what()` doesn't provide enough information. Could you add something like `Can't parse the delimiters`?.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+      for (int i = columnstartindex; i <= columnendindex; i++)
+      {
+        dimension.insert(i);
+      }
```
The braces are redundant.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+      {
+        dimension.insert(std::stoi(temp_dimension[i]));
+      }
+      catch (const std::exception& e)
```
Do we need to catch `std::bad_alloc` as well (`std::set::insert` could throw it)? It has nothing in common with the delimiter format.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+      }
+      catch (const std::exception& e)
+      {
+        Log::Fatal << e.what() << std::endl;      
```
Could you add some information about the error? Something like `Can't parse the delimiter. Usage:...`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+  while (std::getline(fin, line))
+  {
+    stringstream streamLine(line);
+    dataset.push_back(std::vector<std::string>());
```
I think `dataset.emplace_back();` should work.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+  if (CLI::HasParam("column_delimiter"))
+  {
+    column_delimiter = CLI::GetParam<std::string>("column_delimiter");
+    // Allow only 3 delimeter.
```
Pedantic issue: `3 delimiters`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+    {
+      if (col < dimension.size())
+      {
+        if (*dimensionIterator == j)
```
This will never happen if `dimension` contains negative values since `*dimension.begin() < 0`.

jeffin143 replied:
dimension is of type `<size_t>` , how will negative value be there  ?

lozhnikov replied:
Yes, I agree.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+  {
+    size_t col = 0;
+    dimensionIterator = dimension.begin();
+    for (size_t j = 0; j < dataset[i].size(); j++)
```
You don't need the complete enumeration, you can iterate over `std::set`. It's quite easy:
```c++
for (size_t column : dimension)
{
  ...
}
```
Don't forget to verify that the `dimension` set doesn't have invalid values.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+        if (*dimensionIterator == j)
+        {
+          // insert into input
+          input[col].push_back(dataset[i][j]);
```
I think I have already written it, could you find something more meaningful than `input`?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+    }
+    const std::string delimiter = CLI::GetParam<std::string>
+        ("delimiter");
+    for (size_t i   = 0; i < input.size(); i++)
```
Why don't you use `range-based-for` loops?
```c++
for (std::string& line : input)
{
  obj.RemoveStopWords(line, stopwords,
          data::SplitByChar(delimiter));
}
```

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+    {
+      if (col < dimension.size() && *dimensionIterator == j)
+      {
+        fout<<input[col][i]<<column_delimiter;
```
This will never happen if `dimension` has negative values. And you needn't write the delimiter if this is the last column.
I think this line may lead to segmentation fault since you didn't verify that `input[col]` has `dataset[i].size()` elements.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,266 @@
+using namespace arma;
+using namespace std;
+
+static void mlpackMain()
```
The function has about 200 lines. I guess we should divide it into some pieces.

jeffin143 replied:
I am not sure, what does this mean,Should I separate the files, or create functions ?

lozhnikov replied:
I think it's a good idea to divide `mlpackMain()` into a number of smaller functions.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,126 @@
+  {
+    for (auto& str : input)
+    {
+      std::string result = "";
```
Again, `=""` is not needed.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,126 @@
+      while (!token.empty())
+      {
+        if (stopwords.find(token) == stopwords.end())
+          result += (result == "") ? std::string(token) : " " + 
```
```
result += (result == "") ? std::string(token) : " " + std::string(token);
```

The comparison is not needed. The first token has been written outside of the loop.
It's enough to write

```
result += " " + std::string(token);
```

jeffin143 replied:
Take the case of `isn't Mlpack great`

If the first token is a Stop word then result will be empty and then it will have a space in front `result += " " + std::string(token)` , and ans will be ` Mlpack great`

lozhnikov replied:
Yes, you're right.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,267 @@
+      }
+      else
+      {
+        Log::Warn << "column_delimiter not specified, taking default value \n";
```
Looks like you didn't assign the default value to `column_delimiter`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,267 @@
+      try
+      {
+        // Has a range include, something of type a-b.
+        columnstartindex = std::stoi(temp_dimension[i].substr(0, found));
```
`std::stoi()` doesn't throw an exception if the input contains characters other than digits. For example
`std::stoi("1234 abcde")` returns `1234`. I think we should check each character with `std::isdigit()`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,267 @@
+  std::set<size_t>dimension;
+  int columnstartindex, columnendindex;
+  size_t found;
+  for (size_t i = 0; i < temp_dimension.size(); i++)
```
I think the whole loop can be moved to a separate function that returns `std::set<size_t>`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,267 @@
+  std::string line, word;
+  std::vector<std::vector<std::string>> dataset;
+  boost::string_view token, copy;
+  while (std::getline(fin, line))
```
This loop can be moved to a separate function that accepts `filename` and returns `dataset`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,267 @@
+  }
+  const std::string outputFilename = CLI::GetParam<std::string>("preprocess"
+      "_dataset");
+  ofstream fout;
```
The whole code below can be moved to a separate function.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,267 @@
+    dimensionIterator = dimension.begin();
+    for (size_t j = 0 ; j < dataset[i].size(); j++)
+    {
+      if (col < dimension.size() && *dimensionIterator == j)
```
If you replace `std::set<size_t>` by `std::unordered_set<size_t>` you can simplify the comparison:
```
if (dimension.find(j) != dimension.end())`
```
The complexity of `std::unordered_set::find()` is O(1).

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,126 @@
+        result += std::string(token);
+      if (!token.empty())
+        token = tokenizer(strView);
+      std::cout<<"here"<<result<<"opt \n";
```
`std::cout` is not needed here.

jeffin143 replied:
For debugging purpose, I will remove it

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,126 @@
+      }
+      str = std::move(result);
+    }
+  }
```
I didn't take into account that the first token could be a stop word. I think the optimal implementation is the following:
```
token = tokenizer(strView);

while(!token.empty())
{
  if (stopwords.find(token) == stopwords.end())
  {
    result += result.empty() ? std::string(token) : 
                               " " + std::string(token);
  }

  token = tokenizer(strView);
}
```

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+            temp_dimension[i].length()));        
+        columnstartindex = std::stoi(temp_dimension[i].substr(0, found));
+        columnendindex = std::stoi(temp_dimension[i].substr(found+1,
+            temp_dimension[i].length()));
```
You use `substr()` two times in order to get the same substring. Perhaps the compiler can optimize it out. I suggest to save the result to a variable.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+}
+
+void getColumn(const std::vector<std::string>& temp_dimension,
+               std::set<size_t>& dimension)
```
- We discussed that the output variable should be an unordered set since it supports constant time search while `std::set` provides logarithmic search.
- I think it is more natural to return the output rather than pass it as an argument i.e.
  `unordered_set<size_t> GetColumn(...)`
- The container contains a number of dimensions rather than a dimension. I suggest to use the name `dimensions`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+  }
+}
+
+void writeOutput(const std::string& outputFilename,
```
See the comment above.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+                 const std::set<size_t>& dimension)
+{
+  ofstream fout;
+  fout.open(outputFilename, ios::trunc);
```
You can pass these arguments to the constructor.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+    dimensionIterator = dimension.begin();
+    for (size_t j = 0 ; j < dataset[i].size(); j++)
+    {
+      if (col < dimension.size() && *dimensionIterator == j)
```
Again, It's too complecated. You can use `unordered_set` instead of `set`:
```
if (dimenstions.find(j) != dimensions.end())
```

The complexity won't change.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+  // Extracting the filename
+  const std::string filename = CLI::GetParam<std::string>("actual_dataset");
+  // This is very dangerous, Let's add a check tommorrow.
+  std::string column_delimiter;
```
We should use camel case here,

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+      CLI::GetParam<std::vector<std::string> >("dimension");
+  std::set<size_t> dimension;
+  getColumn(temp_dimension, dimension);
+  std::vector<std::vector<std::string>> dataset;
```
There is no need to write `std::`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+  }
+  // Handling Dimension vector
+  std::vector<std::string> temp_dimension =
+      CLI::GetParam<std::vector<std::string> >("dimension");
```
There is no need to write `std::`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+  std::vector<std::vector<std::string>> dataset;
+  createDataset(dataset, filename, column_delimiter);
+  // Preparing the input dataset on which string manipulation has to be done.
+  std::vector<std::vector<std::string>> nonNumericInput(dimension.size());
```
The same comment.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+    {
+      if (col < dimension.size())
+      {
+        if (*dimensionIterator == j)
```
Again, it's too complicated. I suggest to use `std::unordered_set` and `find()`.

jeffin143 replied:
Suppose we have 50 odd columns,
And the set has column 1,6,7 and then

I can break of after 7 th column
So potentially it's row*7

Where in case of unordered_map , I will have to iterate over the column full

I guess 
I can just iterate over the coulmn and insert then value rather than checking 

So I can do iterate over the set 
And just insert 
`nonNumericImput[col].push_back(dataset [I][set_elem]

jeffin143 replied:
I will use unordered_map :)

lozhnikov replied:
You don't need to iterate over the set. You have already iterated over the whole column set:
```c++
for (size_t j = 0; j < dataset[i].size(); j++)
```

lozhnikov replied:
I see. Probably its a good idea to store the column index and the column data inside the same container:
```c++
vector<pair<size_t, vector<string>>> nonNumericInput;
```
or
```c++
unordered_map<size_t, vector<string>> nonNumericInput;
```
In that case you needn't iterate over the whole column set.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+    }
+    ifstream stopwordfile;
+    // Open an existing file
+    const std::string stopwordfilename =
```
We should use camel case for all variables.

jeffin143 commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,301 @@
+{
+  for (auto& i : column)
+    if (!std::isdigit(i))
+    Log::Fatal << "Characters encountered, Please give only digits";
```
you mean, istead of running for loop in function, run it outside and let the function have only check ?

lozhnikov replied:
No, I mean the following:
```c++
for (auto& i : column)
{
  if (!std::isdigit(i))
    return false;
}
```

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+using namespace std;
+
+static vector<vector<string>> CreateDataset(const string& filename,
+                                            const char& columnDelimiter)
```
`columnDelimiter` shouldn't be a reference.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+        string subStringEnd = tempDimension[i].substr(found+1,
+            tempDimension[i].length());
+        IsNumber(subStringStart);
+        IsNumber(subStringEnd);        
```
You forgot to check the return value. You should call `Log::Fatal` if `IsNumber()` returns `false`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+    {
+      try
+      {
+        IsNumber(tempDimension[i]);
```
You should call `Log::Fatal` if the function returns `false`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+                 const vector<vector<string>>& dataset,
+                 const unordered_map<size_t, vector<string>>& nonNumericInput,
+                 const string& columnDelimiter,
+                 const unordered_set<size_t>& dimensions)
```
Could you fix the indent?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+      if (dimensions.find(j) != dimensions.end())
+      {
+        if (j + 1 < dataset[i].size())
+          fout<<nonNumericInput.at(j)[i]<<columnDelimiter;
```
According to the style guide we should add spaces around each binary operator.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+        if (j + 1 < dataset[i].size())
+          fout<<nonNumericInput.at(j)[i]<<columnDelimiter;
+        else
+          fout<<nonNumericInput.at(j)[i];
```
According to the style guide we should add spaces around each binary operator.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+      else
+      {
+        if (j + 1 < dataset[i].size())
+          fout<<dataset[i][j]<<columnDelimiter;
```
According to the style guide we should add spaces around each binary operator.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+        if (j + 1 < dataset[i].size())
+          fout<<dataset[i][j]<<columnDelimiter;
+        else
+          fout<<dataset[i][j];
```
According to the style guide we should add spaces around each binary operator.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+          fout<<dataset[i][j];
+      }
+    }
+    fout<<"\n";
```
According to the style guide we should add spaces around each binary operator.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+  {
+    columnDelimiter = CLI::GetParam<string>("columnDelimiter");
+    // Allow only 3 delimiters.
+    RequireParamValue<string>("columnDelimiter", [](string del)
```
Probably `del` should be a const reference.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+      {
+        columnDelimiter = "\t";
+        Log::Warn << "Found tsv or txt Extension, taking \\t as"
+        "columnDelimiter. \n";
```
There is no space between `as` and `columnDelimiter`.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+  // File stream.
+  ifstream fin(filename);
+  if (!fin.is_open())
+    Log::Fatal << "Unable to open input file \n";
```
I think each log message should end with `endl` since `\n` doesn't flush the buffer. I didn't point out the same issues below. Could you fix them as well?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,273 @@
+  for (size_t i = 0; i < dataset.size(); i++)
+  {
+    for (auto& datasetCol : dimensions)
+      nonNumericInput[datasetCol].push_back(dataset[i][datasetCol]);
```
I guess you are not going to use `dataset[i][datasetCol]` anymore. In that case it's better to move it to the vector than copy.
```
nonNumericInput[datasetCol].push_back(move(dataset[i][datasetCol]));
```

jeffin143 replied:
I need it for writting results back to the ouptut file, since preprocessed data is in nonNumericInput and dataset has numeric input.

lozhnikov replied:
I mean you don't need to store the columns that you move to `nonNumericInput` twice.

jeffin143 commented:
@lozhnikov , I addressed all your comments, can you PTAL, also I have to write test for The binding now, once if you are fine with the binding i will write its test.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,123 @@
+/**
+ * @file string_cleaning.hpp
+ * @author jeffin sam
+ *
```
We should add the description.
```
Definition of the StringCleaning class.
```

lozhnikov replied:
@jeffin143 Why did you mark this issue as resolved? I don't see the fix.

jeffin143 replied:
https://github.com/mlpack/mlpack/commit/03943e97778da97e822a17bd6a1885b90682c1b0

I did add, sorry is it not correct

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,123 @@
+namespace mlpack {
+namespace data {
+
+class StringCleaning
```
Could you add a brief class description?

lozhnikov replied:
The same problem.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,123 @@
+  /**
+  * Function to remove punctuation from a given vector of strings.
+  *
+  * @param input Vector of strings
```
The second parameter isn't described.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,123 @@
+  * function. Its return value should be of bool type.
+  *
+  * Definiation of function should be of type
+  * bool fn(const char& c)
```
Probably the reference is redundant.

lozhnikov commented on src/mlpack/core/data/string_cleaning.hpp:
```diff
@@ -0,0 +1,123 @@
+      boost::algorithm::to_lower(str);
+  }
+  /**
+  * Function to convert given vector of strings to Upper case.
```
Small typo: `upper`

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+PROGRAM_INFO("preprocess_string",
+    // Short description.
+    "A utility to preprocess string data. This utility can remove stopwords, "
+    "punctuation and convert to lowercase.",
```
and uppercase.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+    "t");
+PARAM_STRING_IN_REQ("preprocess_dataset", "File containing the preprocess "
+    "dataset.", "o");
+PARAM_STRING_IN("columnDelimiter", "delimeter used to seperate Column in files"
```
CLI arguments are snake case.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+using namespace std;
+
+static vector<vector<string>> CreateDataset(const string& filename,
+                                            char& columnDelimiter)
```
Why is `columnDelimiter` a reference?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+using namespace arma;
+using namespace std;
+
+static vector<vector<string>> CreateDataset(const string& filename,
```
Could you add a brief description before each function?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+  {
+    stringstream streamLine(line);
+    dataset.emplace_back();
+    // delimeter[0] becase the standard function accepts char as input.
```
The comment is out of date.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+
+static bool IsNumber(const string& column)
+{
+  for (auto& i : column)
```
It's rather odd to pass `char` by reference.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+}
+
+static unordered_set<size_t> GetColumnIndices(const
+                                              vector<string>& tempDimension)
```
The line break after the `const` modifier looks terrible.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+      {
+        // Has a range include, something of type a-b.
+        string subStringStart = tempDimension[i].substr(0, found);
+        string subStringEnd = tempDimension[i].substr(found+1,
```
Again,
```
found + 1
```

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+      CLI::GetParam<vector<string> >("dimension");
+  unordered_set<size_t> dimensions = GetColumnIndices(tempDimension);
+  vector<vector<string>> dataset = CreateDataset(filename, columnDelimiter[0]);
+  for (auto colIndex : dimensions)
```
Could you add braces here?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+      stopwords.insert(originalword.back());
+    }
+    const string delimiter = CLI::GetParam<string>
+        ("delimiter");
```
I think the line break is redundant.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+      Log::Fatal << "Please provide a file for stopwords." << endl;
+    // Open an existing file
+    const string stopWordFilename =
+        CLI::GetParam<string>("stopwordsfile");
```
Looks like the line break is redundant.

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+    }
+    const string delimiter = CLI::GetParam<string>
+        ("delimiter");
+    for (auto& datasetCol : dimensions)
```
Could you add braces here?

lozhnikov commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,284 @@
+  if (CLI::HasParam("lowercase"))
+  {
+    for (auto& datasetCol : dimensions)
+      obj.LowerCase(nonNumericInput[datasetCol]);
```
I think the following code is faster than the present code:
```
for (auto& column : nonNumericInput)
  obj.LowerCase(column->second);
```
Could you fix the rest of the code accordingly?

lozhnikov commented:
@jeffin143 You pointed out the whole latest review as resolved. In fact you fixed none of the issues. Did you forget to push the code?

jeffin143 commented:
> @jeffin143 You pointed out the whole latest review as resolved. In fact you fixed none of the issues. Did you forget to push the code?

https://github.com/mlpack/mlpack/commit/03943e97778da97e822a17bd6a1885b90682c1b0 
I did commit it an hour ago , I guess I resolved some of those

lozhnikov commented:
> I did commit it an hour ago , I guess I resolved some of those

Yes, I see now.

lozhnikov commented on src/mlpack/tests/main_tests/preprocess_string_test.cpp:
```diff
@@ -0,0 +1,431 @@
+    "mlpack. is!' a fast! machine learning.", "gsoc. is!' great!." }  
+  };
+  CreateFile(stringLowerInput,"preprocess_string_lower_test.txt","\t" );
+  SetInputParam("actual_dataset", (std::string) "string_test.txt");
```
Could you remove all C-style conversions?
I mean the following:
```c++
(std::string) "string_test.txt"
```
The following code should work:
```c++
SetInputParam<std::string>("actual_dataset", "string_test.txt");
```

lozhnikov commented:
I think the PR is almost ready. I am going to write benchmarks and fix the implementation (if needed). Probably I'll open a PR.

zoq commented:
Should we whitelist the style errors or can we solve most of them, just skimmed over a couple and I think we could resolve some of them.

zoq commented on src/mlpack/tests/string_cleaning_test.cpp:
```diff
@@ -0,0 +1,99 @@
+}
+
+/**
+ * Test for removal of char
```
Picky style comment missing stop at the end.

zoq commented on src/mlpack/tests/string_cleaning_test.cpp:
```diff
@@ -0,0 +1,99 @@
+}
+
+/**
+ * Test for removal of punctuation
```
See comment below.

zoq commented on src/mlpack/tests/string_cleaning_test.cpp:
```diff
@@ -0,0 +1,99 @@
+BOOST_AUTO_TEST_SUITE(StringCleaningTest);
+
+/**
+ * Test for removal of stop words
```
See comment below.

zoq commented on src/mlpack/core/data/tokenizer/char_split.hpp:
```diff
@@ -0,0 +1,47 @@
+  * return a boost::string_view (token).
+  * @param str A string to retriev token from.
+  */
+  boost::string_view operator()(boost::string_view& str) const
```
Looks like ``str`` could be ``const``.

zoq commented on src/mlpack/core/data/tokenizer/split_by_char.hpp:
```diff
@@ -0,0 +1,76 @@
+{
+ public:
+  /**
+  * A constructor to set delimiter.
```
This is off by one space.

zoq commented on src/mlpack/core/data/tokenizer/split_by_char.hpp:
```diff
@@ -0,0 +1,76 @@
+  {
+    this->delimiter = delimiter;
+  }
+  /**
```
Might be a good idea to add a new line here to improve the readability.

zoq commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,307 @@
+        if (!IsNumber(subStringStart) || !IsNumber(subStringEnd))
+        {
+          Log::Fatal << "Dimension value accepts only digits, characters"
+              " encountered, please recheck" << endl;
```
Missing stop at the end of the error message.

zoq commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,307 @@
+        if (!IsNumber(tempDimension[i]))
+        {
+          Log::Fatal << "Dimension value accepts only digits, characters"
+              " encountered, please recheck" << endl;
```
See comment above.

zoq commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,307 @@
+      }
+      catch (const exception& e)
+      {
+        Log::Fatal << "Dimension value not appropriate " << endl;
```
See comment above.

zoq commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,307 @@
+ * @param dataset The actual dataset which was read from file.
+ * @param nonNumericInput The preproccessed data
+ * @param columnDelimiter Delimiter used to separate columns of the file.
+ * @param dimesnions Dimesnion which we non numeric or of type string.
```
This should be "dimensions"; also not sure I get the description.

zoq commented on src/mlpack/methods/preprocess/preprocess_string_main.cpp:
```diff
@@ -0,0 +1,307 @@
+  // Parse command line options.
+  // Extracting the filename
+  const string filename = CLI::GetParam<string>("actual_dataset");
+  // This is very dangerous, Let's add a check tommorrow.
```
What do we have to check here?

jeffin143 commented:
@lozhnikov should we remove clu binding and merge this too ? And then I work on the cli again 

Or do I have to re write this

lozhnikov commented:
> @lozhnikov should we remove clu binding and merge this too ? And then I work on the cli again
> 
> Or do I have to re write this

I'll open a PR on the basis of this one.

jeffin143 commented:
@lozhnikov friendly ping
